internal/graph/graph_test.go

```diff
@@ -0,0 +1,546 @@
+										typesystem.RelationReference("user", ""),
+										typesystem.RelationReference("group", "member"),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			targetUserRef:   typesystem.RelationReference("user", ""),
+			targetObjectRef: typesystem.RelationReference("document", "viewer"),
+			expectedIngresses: []*RelationshipIngress{
+				{
+					Type:    DirectIngress,
+					Ingress: typesystem.RelationReference("document", "viewer"),
+				},
+				{
+					Type:    DirectIngress,
+					Ingress: typesystem.RelationReference("document", "editor"),
+				},
+				{
+					Type:    DirectIngress,
+					Ingress: typesystem.RelationReference("group", "member"),
+				},
+			},
+		},
+		{
+			name: "Circular reference",
+			model: &openfgapb.AuthorizationModel{
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "user",
+					},
+					{
+						Type: "team",
+						Relations: map[string]*openfgapb.Userset{
+							"member": typesystem.This(),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"member": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										typesystem.RelationReference("group", "member"),
+									},
+								},
+							},
+						},
+					},
+
+					{
+						Type: "group",
+						Relations: map[string]*openfgapb.Userset{
+							"member": typesystem.This(),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"member": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										typesystem.RelationReference("user", ""),
+										typesystem.RelationReference("team", "member"),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			targetUserRef:   typesystem.RelationReference("user", ""),
+			targetObjectRef: typesystem.RelationReference("team", "member"),
+			expectedIngresses: []*RelationshipIngress{
+				{
+					Type:    DirectIngress,
+					Ingress: typesystem.RelationReference("group", "member"),
+				},
+			},
+		},
+		{
+			name: "No graph relationship connectivity",
+			model: &openfgapb.AuthorizationModel{
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "user",
+					},
+					{
+						Type: "team",
+						Relations: map[string]*openfgapb.Userset{
+							"member": typesystem.This(),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"member": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										typesystem.RelationReference("team", "member"),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			targetUserRef:     typesystem.RelationReference("user", ""),
+			targetObjectRef:   typesystem.RelationReference("team", "member"),
+			expectedIngresses: []*RelationshipIngress{},
+		},
+		{
+			name: "",
```

Comment:
Not sure the best name for this? Any recommendations 😄 ?

Reply:
it's hard to come up with a name unless you understand the model, and in its current form the model is hard to understand... @craigpastro can you put a comment on each test with the DSL representation?